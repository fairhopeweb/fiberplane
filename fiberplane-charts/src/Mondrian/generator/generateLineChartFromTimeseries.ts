import type {
  AbstractChart,
  Axis,
  Point,
  Shape,
  ShapeList,
  TimeseriesSourceData,
} from "../types";
import {
  attachSuggestionsToXAxis,
  calculateSmallestTimeInterval,
  calculateYAxisRange,
  createMetricBuckets,
  extendMinMax,
  getInitialMinMax,
  getTimeFromTimestamp,
  getXAxisFromTimeRange,
  normalizeAlongLinearAxis,
  splitIntoContinuousLines,
} from "./utils";
import { identity } from "../../utils";
import type { Metric, Timeseries } from "../../providerTypes";
import type { MinMax } from "./types";

export function generateLineChartFromTimeseries(
  input: TimeseriesSourceData,
): AbstractChart<Timeseries, Metric> {
  const buckets = createMetricBuckets(
    input.timeseriesData,
    (maybeMinMax: MinMax | undefined, value) =>
      maybeMinMax ? extendMinMax(maybeMinMax, value) : getInitialMinMax(value),
  );

  const xAxis = getXAxisFromTimeRange(input.timeRange);
  const yAxis = calculateYAxisRange(buckets, identity);

  const interval = calculateSmallestTimeInterval(buckets);
  if (interval) {
    attachSuggestionsToXAxis(xAxis, buckets, interval);
  }

  const shapeLists: Array<ShapeList<Timeseries, Metric>> =
    input.timeseriesData.map((timeseries) => ({
      shapes: timeseries.visible
        ? getShapes(timeseries.metrics, xAxis, yAxis, interval)
        : [],
      source: timeseries,
    }));

  return { shapeLists, xAxis, yAxis };
}

function getShapes(
  metrics: Array<Metric>,
  xAxis: Axis,
  yAxis: Axis,
  interval: number | null,
): Array<Shape<Metric>> {
  switch (metrics.length) {
    case 0:
      return [];
    case 1: {
      const metric = metrics[0];
      return Number.isNaN(metric.value)
        ? []
        : [
            {
              type: "point",
              ...getPointForMetric(metric, xAxis, yAxis),
            },
          ];
    }
    default: {
      const lines = splitIntoContinuousLines(metrics, interval ?? undefined);
      return lines.map((line) =>
        // If the line only containes one metric value, render it as a point
        // Otherwise, render a line
        line.length === 1
          ? {
              type: "point",
              ...getPointForMetric(line[0], xAxis, yAxis),
            }
          : {
              type: "line",
              points: line.map((metric) =>
                getPointForMetric(metric, xAxis, yAxis),
              ),
            },
      );
    }
  }
}

function getPointForMetric(
  metric: Metric,
  xAxis: Axis,
  yAxis: Axis,
): Point<Metric> {
  const time = getTimeFromTimestamp(metric.time);

  return {
    x: normalizeAlongLinearAxis(time, xAxis),
    y: normalizeAlongLinearAxis(metric.value, yAxis),
    source: metric,
  };
}
