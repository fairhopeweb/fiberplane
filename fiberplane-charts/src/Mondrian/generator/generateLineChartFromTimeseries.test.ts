import type { Metric, Timeseries } from "../../providerTypes";
import { generateLineChartFromTimeseries } from "./generateLineChartFromTimeseries";

test("it generates a line chart", () => {
  const timeseries: Timeseries = {
    name: "dummy_data",
    labels: {},
    attributes: {},
    resource: {},
    metrics: [
      getMetricAtMinute(0, 10),
      getMetricAtMinute(1, 15),
      getMetricAtMinute(2, 20),
      getMetricAtMinute(4, 25),
      getMetricAtMinute(5, 30),
      getMetricAtMinute(6, 25),
      getMetricAtMinute(7, 20),
      getMetricAtMinute(8, 15),
      getMetricAtMinute(9, 10),
    ],
    visible: true,
  };

  expect(
    generateLineChartFromTimeseries({
      graphType: "line",
      stackingType: "none",
      timeRange: {
        from: getDateAtMinute(0).toISOString(),
        to: getDateAtMinute(10).toISOString(),
      },
      timeseriesData: [timeseries],
      additionalValues: [40],
    }),
  ).toMatchSnapshot();
});

function getMetricAtMinute(minute: number, value: number): Metric {
  return {
    time: getDateAtMinute(minute).toISOString(),
    value,
    attributes: {},
    resource: {},
  };
}

function getDateAtMinute(minute: number): Date {
  return new Date(
    `2023-07-18T16:${minute.toString().padStart(2, "0")}:00.000Z`,
  );
}
