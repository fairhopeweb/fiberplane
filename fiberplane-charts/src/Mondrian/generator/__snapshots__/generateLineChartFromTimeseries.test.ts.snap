// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates a line chart 1`] = `
{
  "shapeLists": [
    {
      "shapes": [
        {
          "points": [
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:00:00.000Z",
                "value": 10,
              },
              "x": 0,
              "y": 0.045454545454545456,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:01:00.000Z",
                "value": 15,
              },
              "x": 0.1,
              "y": 0.2727272727272727,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:02:00.000Z",
                "value": 20,
              },
              "x": 0.2,
              "y": 0.5,
            },
          ],
          "type": "line",
        },
        {
          "points": [
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:04:00.000Z",
                "value": 25,
              },
              "x": 0.4,
              "y": 0.7272727272727273,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:05:00.000Z",
                "value": 30,
              },
              "x": 0.5,
              "y": 0.9545454545454546,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:06:00.000Z",
                "value": 25,
              },
              "x": 0.6,
              "y": 0.7272727272727273,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:07:00.000Z",
                "value": 20,
              },
              "x": 0.7,
              "y": 0.5,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:08:00.000Z",
                "value": 15,
              },
              "x": 0.8,
              "y": 0.2727272727272727,
            },
            {
              "source": {
                "attributes": {},
                "resource": {},
                "time": "2023-07-18T16:09:00.000Z",
                "value": 10,
              },
              "x": 0.9,
              "y": 0.045454545454545456,
            },
          ],
          "type": "line",
        },
      ],
      "source": {
        "attributes": {},
        "labels": {},
        "metrics": [
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 10,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 10,
          },
        ],
        "name": "dummy_data",
        "resource": {},
        "visible": true,
      },
    },
  ],
  "xAxis": {
    "maxValue": 1689696600,
    "minValue": 1689696000,
    "tickSuggestions": [
      1689696000,
      1689696060,
      1689696120,
      1689696180,
      1689696240,
      1689696300,
      1689696360,
      1689696420,
      1689696480,
      1689696540,
    ],
  },
  "yAxis": {
    "maxValue": 31,
    "minValue": 9,
  },
}
`;
